File: cpp/organism/animal/species/Wolf.cpp
//
// Created by Filip on 12/04/2025.
//

#include "Wolf.h"

Wolf::Wolf(Point point, World &world)
    : Animal(point, Constants::Animal::Wolf::Symbol,
             Constants::Animal::Wolf::Strength,
             Constants::Animal::Wolf::Initiative, world) {}

void Wolf::reproduce() {
  Point newPosition = world.getFreeSpace(position);
  if (newPosition == Point(-1, -1)) {
    return;
  }

  world.queueOrganismAddition(new Wolf(newPosition, world));
}

File: java/src/organism/animal/species/Wolf.java
package organism.animal.species;

import constants.Constants;
import logger.EventLogger;
import organism.Organism;
import organism.animal.Animal;
import world.World;

public class Wolf extends Animal {

    public Wolf(int x, int y, World world) {
        super(x, y, Constants.Animal.Wolf.Symbol, Constants.Animal.Wolf.Strength, Constants.Animal.Wolf.Initiative, world);
    }

    @Override
    protected void reproduce() {
        int[] newPosition = world.findNearestFree(this.getX(), this.getY());
        if (newPosition != null) {
            // Create a new organism of the same type
            Animal offspring = new Wolf(newPosition[0], newPosition[1], world);
            world.addOrganism(offspring);
            EventLogger.getInstance().log("Organism " + this.getClass().getSimpleName() + " reproduced at (" + newPosition[0] + ", " + newPosition[1] + ")");
        }
    }

    @Override
    public boolean collision(Organism other) {
        return super.collision(other);
    }
}